pkg/cli/options.go

```diff
@@ -139,3 +150,150 @@ func WithCompletion() Option {
+		logrus.Errorf("Host not supported: %s", host)
+	}
+
+	return pluginsRoot
+}
+
+// DiscoverExternalPlugins discovers the external plugins in the plugins root directory and adds them to external.Plugin.
+func DiscoverExternalPlugins() (ps []plugin.Plugin, err error) {
+	pluginsRoot := setPluginsRootBasedonHost()
+
+	rootInfo, err := os.Stat(pluginsRoot)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			logrus.Debugf("External plugins dir %q does not exist, skipping external plugin parsing", pluginsRoot)
+			return nil, nil
+		}
+		return nil, err
+	}
+	if !rootInfo.IsDir() {
+		logrus.Debugf("External plugins path %q is not a directory, skipping external plugin parsing", pluginsRoot)
+		return nil, nil
+	}
+
+	infos, err := ioutil.ReadDir(pluginsRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if !info.IsDir() {
+			logrus.Debugf("%q is not a directory so skipping parsing", info.Name())
+			continue
+		}
+
+		subinfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		for _, subinfo := range subinfos {
+			pluginInfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name(), subinfo.Name()))
+			if err != nil {
+				return nil, err
+			}
+			for _, pluginInfo := range pluginInfos {
+				// find the executable that matches the same name as info.Name().
+				// if no match is found, compare the external plugin string name before dot and match it with info.Name() which is the external plugin root dir.
+				// for example: sample.sh --> sample, externalplugin.py --> externalplugin
+				trimmedPluginName := strings.Split(pluginInfo.Name(), ".")
+				if len(trimmedPluginName) == 0 {
+					return nil, fmt.Errorf("Invalid plugin name found %q", pluginInfo.Name())
+				}
+
+				if pluginInfo.Name() == info.Name() || trimmedPluginName[0] == info.Name() {
+					// check whether the external plugin is an executable.
+					if !isPluginExectuable(pluginInfo.Mode()) {
+						return nil, fmt.Errorf("External plugin %s found in path is not an executable", pluginInfo.Name())
+					}
+
+					_, args := parseExternalPluginArgs()
+
+					ep := external.Plugin{
+						PName: info.Name(),
+						Path:  filepath.Join(pluginsRoot, info.Name(), subinfo.Name(), pluginInfo.Name()),
+						// todo(rashmigottipati): external.Plugin must satisfy the Plugin interface,
+						// so adding the supported project version as a placeholder value for now.
+						PSupportedProjectVersions: []config.Version{cfgv2.Version, cfgv3.Version},
+						Args:                      args,
+					}
+
+					if err := ep.PVersion.Parse(subinfo.Name()); err != nil {
+						return nil, err
+					}
+
+					logrus.Printf("Adding external plugin: %s", ep.Name())
+
+					ps = append(ps, ep)
+
+				}
+			}
+		}
+
+	}
+	return ps, nil
+}
+
+// isPluginExectuable checks if a plugin is an executable based on the bitmask and returns true or false.
+func isPluginExectuable(mode os.FileMode) bool {
```

Comment:
I wouldn't try to check for whether a file is executable or not. Let `ExternalPlugin.Run()` handle this error, since it will be more informative.

Reply:
@estroz why? Seems like we want to ensure that the file we deem is the "executable for the plugin" is correct when we read it. In my head since we are already scanning files to see which one needs to be associated with the plugin that we can deem whether it will even work down the line. Otherwise, it will get down to the Run step and fail. 

Maybe I don't know what the error will look like yet. I just feel like it would be better to fail early than later.

Reply:
That was kind of my hunch too. To fail early during discovery itself and not even add it to the list of plugins. 

Will check if `ExternalPlugin.Run()` handles this check.

Reply:
If discovery skips a plugin file because it's not executable, then the CLI will report that the plugin is not found, which is confusing imo. Instead, just treat the plugin as found so the CLI can match passed `--plugins` values, then fail during execution with a more informative error.

Reply:
Just tested on a random file that matches the same name as the plugin name passed in the CLI but that's not an executable. Either way, it would fail, but have to figure out what error messages are more informative. 

Without the check seems more informative as it shows the entire plugin path. Any thoughts @estroz @jmrodri ? 

With the check:
▶ ../bin/kubebuilder init --plugins random
2021/09/24 13:30:48 External plugin random found in path is not an executable

Without the check: 
 ▶ ../bin/kubebuilder init --plugins random
INFO[0000] Adding external plugin: random
Error: failed to initialize project: unable to scaffold with "random/v1": fork/exec /Users/rashmigottipati/Library/ApplicationSupport/kubebuilder/plugins/random/v1/random: permission denied
Usage:
  kubebuilder init [flags]

Examples:
  # Help for initializing a project with version "2"
  kubebuilder init --project-version="2" -h

  # Help for initializing a project with version "3"
  kubebuilder init --project-version="3" -h

Flags:
  -h, --help                     help for init
      --project-version string   project version (default "3")

Global Flags:
      --plugins strings   plugin keys to be used for this subcommand execution

2021/09/24 13:31:07 failed to initialize project: unable to scaffold with "random/v1": fork/exec /Users/rashmigottipati/Library/ApplicationSupport/kubebuilder/plugins/random/v1/random: permission denied

Reply:
I think the first one is better actually. The check needs to be improved though. Comments below.

Reply:
Sounds good.